extensions/amp-ad-network-base/0.1/amp-ad-type-defs.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @typedef {{
+      minifiedCreative: string,
+      customElementExtensions: !Array<string>,
+      customStylesheets: !Array<{href: string}>,
+      images: (Array<string>|undefined),
+    }} */
+export let CreativeMetaDataDef;
+
+/** @typedef {{
+      templateUrl: string,
+      data: (JsonObject|undefined),
+      analytics: (JsonObject|undefined),
+    }} */
+export let AmpTemplateCreativeDef;
+
+/** @typedef {{
+      rawCreativeBytes: !ArrayBuffer,
+      additionalContextMetadata: !JsonObject,
+      sentinel: string,
+    }} */
+export let CrossDomainDataDef;
+
+/** @typedef {{width: string, height: string}} */
+export let LayoutInfoDef;
+
+/** @enum {string} */
+export const FailureType = {
+  REQUEST_ERROR: 'REQUEST_ERROR',
+  INVALID_RESPONSE: 'INVALID_RESPONSE',
+  EMPTY_RESPONSE: 'EMPTY_RESPONSE',
+  VALIDATOR_ERROR: 'VALIDATOR_ERROR',
+  RENDERER_ERROR: 'RENDERER_ERROR',
+};
+
+/** @enum {string} */
+export const RecoveryModeType = {
+  COLLAPSE: 'COLLAPSE',
+  RETRY: 'RETRY',
+};
+
+/** @enum {string} */
+export const ValidatorResult = {
+  AMP: 'AMP',
+  NON_AMP: 'NON_AMP',
+};
+
+/**
+ * @abstract
+ */
+export class Renderer {
+  /**
+   * @param {!./amp-ad-context.AmpAdContext} unusedContext
+   * @param {!./amp-ad-network-base.AmpAdNetworkBase} unusedBaseInstance
+   * @return {!Promise<!./amp-ad-context.AmpAdContext>}
+   * @abstract
+   */
+  render(unusedContext, unusedBaseInstance) {}
```

Comment:
why does it take AmpAdNetworkBase as param?

Reply:
Some renderers will need to refer to the base instance's `element` and `win` fields, as well as call functions like `getAmpDoc()`, and `isInViewPort()`. I suppose we can move most of that into AmpAdContext, though I'm not sure how ideal that is with something like `isInViewPort()`.

Reply:
Forgot to mention `applyFillContent()`.

Reply:
- `ampdoc` can be passed in as a param.
- `isInViewport()` is used to set visibility, which should be rather done outside the `render()` call

Reply:
Not every renderer calls `isInViewPort()`, so we'd need to move some rendering logic into the base class to know when to call this method.

Reply:
For individual renderer specific stuff, it should go to `context` object.

We'd not worry about it for now. It's a bit unclear to me why it is needed. We can introduce it when working on the friendlyiframe renderer

Reply:
Okay, nixed this param.